text/0000-named-tracers.md

```diff
@@ -0,0 +1,69 @@
+
+This proposal originates from an `opentelemetry-specification` proposal on [components](https://github.com/open-telemetry/opentelemetry-specification/issues/10), but the term "components" was quickly found to being too generic and possibly misleading, since it is already used frequently in differing contexts and connotations.
+
+The purpose of _Named Tracers_ is the ability to allow for a more fine-grained control of the amount of tracing data being generated by a library or application. Associating a name (e.g. _"io.opentelemetry.contrib.mongodb"_) with a tracer allows for a logical coupling of this tracer and a certain technology or library (like _MongoDB_ in this example).
+
+*Named Tracers* allow enabling and disabling tracing of a specific library, which might be necessary in case:
+* A library causes too much runtime or memory overhead.
+* A library has issues or bugs (e.g. does not call `span.end()` in every case).
+* "Double-Tracing" becomes a problem. This can be the case if a vendor (e.g. Dynatrace) might already have built-in auto-instrumentation for a certain library. In this case the vendor might want to decide (automatically or configuration-based) which of the implementations should be enabled or disabled.
+
+
+## Explanation
+
+From a user perspective, working with *Named Tracers* and *TracerFactories* is very similar to how logging frameworks like [log4j](https://www.slf4j.org/apidocs/org/slf4j/LoggerFactory.html) work. In analogy to requesting Logger objects through LoggerFactories, a tracing library would create specific 'Tracer' objects through a 'TracerFactory'.
+
+```java
+Tracer tracer = OpenTelemetry.Tracing.getTracerFactory().getTracer("io.opentelemetry.contrib.mongodb");
+```
+
+In a way, the `TracerFactory` replaces the global `Tracer` singleton object as a ubiquitous point to request a tracer instance.
+
+
+## Internal details
+
+By providing a TracerFactory and *Named Tracers*, a vendor or OpenTelemetry implementation gains more flexibility in providing tracers and can decide whether or not to produce spans for a specific traced component and even change this behavior at runtime.
+
+In the simplest case, an OpenTelemetry implementation can return a single instance for a requested tracer regardless of the name specified. This could be the case for implementations that do not want/need to enable or disable a tracer.
+
+Alternatively, an implementation can provide different tracer instances per specified tracer name, thus being able to associate this tracer with the component being traced. This allows for the possibility to enable / disable a tracer based on a custom configuration.
```

Comment:
This needs to prescribe the actual behavior. What does it mean to "disable" the tracer for some component that's in the middle of the stack, e.g. layer B in A->B->C? What happens when instrumentation still creates a new span? Is it not going to be exported? what happens to the child of it (layer C), what would it be its parent?

Reply:
This RFC targets the API level rather than the SDK. Hence, questions like tracer verbosity or turning tracers on / off are something the SDK or a vendor-implementation can decide based on e.g. a proprietary configuration. On an API level there would currently be no notion of a configuration.

For example, the spec does not prescribe what happens with the span name either. It (more or less) describes what it means, but it does not set any rules how implementation should process the span name. It is true that it would make sense for the SDK to have some additional config APIs to e.g. discard span operations based on the tracer name, but this is out of scope for this RFC

Reply:
> the spec does not prescribe what happens with the span name either.

That's not true, it does provide guidance of what constitutes a good span name and why (at least opentracing spec did), and that guidance is based on some basic understanding of how span name will be used (e.g. to treat spans with the same name as a class, aggregate across them, etc.) The same applies to all data elements collected through the API, there's some shared understanding why they are useful and worth collecting.

If we have no idea _how_ the tracer name can be used, it doesn't make sense in the API either. The main reason I've seen so far is to "turn instrumentation for different layers on and off", but as I showed in the previous comment, a naive implementation of that doesn't work. So aside from labelling a component (which can be done via semantic convention of the tags and does not require API change), I think we need to see a concrete example of how it can be used.

In contrast, in Java logging APIs, the meaning of the layer name is pretty clear, and even more, the names are expected to be in a specific package-like format that implies a hierarchy.

Reply:
It seems to me that this discussion is stuck somehow. What could we do to bring this RFC forward?

Let me try and sum up the main intention of this RFC:

* Establish a basic mechanism to create multiple instances of tracers that can be treated as separate logical entities (as opposed to a single global instance).
* By giving those entities a name, we later have the possibility to:
  * have specific configuration (on/off, verbosity, ...) for a tracer by identifying it through this name.
  * Matching produced spans to a certain tracer by e.g. setting this name as their "component".

If we want to have that same naming mechanism also for metrics, we could have a similar RFC there? I see how that makes sense.